### PR TITLE
[dynamo] Hide guard_fail_hook behind a flag to improve cache lookup time (+10% DebertaV2)

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100599
* __->__ #100590

For TorchDynamo eager backend, DebertaV2 speedup improves from 0.77x to 0.87x. 

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by voznesenskym — APPROVED** [review]
LGTM, lineage here https://github.com/pytorch/pytorch/pull/98280 (included for posterity, comments, discussions, etc)

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -1095,6 +1095,14 @@ def __call__(self, gm: torch.fx.GraphModule, example_inputs):
                 self.op_count += 1
         return gm.forward
 
+    def __enter__(self):
+        self.old_report_guard_failure = config.report_guard_failures
```
Comment (wconstab) [inline torch/_dynamo/utils.py:1099]: nice solution here. thanks!

**Review by wconstab — APPROVED** [review]
lgtm, thanks for finding a good way to auto-enable!

**Pull request closed without merging**